### DISCUSSION TITLE
Dynamically loadable applications

### DISCUSSION THREAD

**Discussion by rlanders** [mbed-ce/mbed-os#302] [Ideas]
I know there are a few responses to this idea.

1. It can't be done because there is no MMU. (from people who always say things can't be done)
2. Just use the -fpic option. (This maybe could work, but it's likely only part of the answer)
3. How about something like microPython? (I'm not saying this wouldn't work, but what about multiple apps at the same time?)
4. Have you looked at TockOS? (This looks promising, but I would need to learn Rust, and...)

OK, so here's the deal, TockOS does use Position Independent Code, and actually the latest try in the Zephyr project does too. And I was bascially leading when I said -fpic is only part of the answer. What we need are .h files that point to the exposed system functions we want. The problem is that everyone's "thingy" is a bit different, and the functions that you want exposed are likely a bit different.

This leads me to my actual "idea". What if we had a tool that allowed someone to pick which things they wanted to expose, parse the map file, and generate a header file that would allow 3rd party developers to write dynamically loadable applications for the target system. In fact, this is the work that TockOS has done, and they have Tockloader to handle the loading/unloading, and even enable/disabling of the applications. So, the boiled down idea is.... Let's steal what TockOS is doing, and let people write in C/C++ instead of having to learn Rust.

**Comment by multiplemonomials** [comment]
My immediate response to this is... in what situations would this be useful?  In most cases, Mbed is used for small MCUs that are only running one application at a time.  And, (though this project admittedly isn't integrated well with Mbed CE), there is the MCUBoot bootloader that allows downloading & installing application updates.  So, in what scenarios would Mbed OS users want to dynamically download and run a separate application, where it could not be accomplished with a regular firmware update?  (Not criticizing, I honestly do not know what these scenarios would be -- please describe some to me!)
